### DISCUSSION TITLE
Generate array always for specified xml element

### DISCUSSION THREAD

**Discussion by ajit4sbi** [mikefarah/yq#1489] [Q&A]
### Generate array always for specified xml element **"aitem"** even with single set

```
<root>
    <aitem>
        <name>abc</name>
        <value>123</value>
    </aitem>
    <bitem>
        <name>bbbb</name>
        <value>2222</value>
    </bitem>
    <bitem>
        <name>BB</name>
        <value>22</value>
    </bitem>
</root>

```

### Command used 
`yq -p=x -o=j test.xml`

### Command Output

```
 {
  "root": {
    "aitem": {
      "name": "abc",
      "value": "123"
    },
    "bitem": [
      {
        "name": "bbbb",
        "value": "2222"
      },
      {
        "name": "BB",
        "value": "22"
      }
    ]
  }
}
```

### Expected Output

 

```
{
  "root": {
    "aitem": [
    {
      "name": "abc",
      "value": "123"
    }
    ],
    "bitem": [
      {
        "name": "bbbb",
        "value": "2222"
      },
      {
        "name": "BB",
        "value": "22"
      }
    ]
  }
 }
```

**Comment by mikefarah** [comment]
You need to convert 'aitem' to an array (yq knew `bitem` is an array because there are multiple entries of it):

```bash
yq -p=x -o=j '.aitem |= [.]' test.xml
```

**Comment by ajit4sbi** [comment]
@mikefarah 

### Any idea can i create xsd schema file & ask command to do accordingly as per schema file.

### Requirement as below , would need two tasks doing in single command.  

- I would like to specify many ".aitem, .oitem, .sitem" like single set elements to be an array
- convert all numbers/floats/booleans to be without double quoted




### Below is the output, missed marking array at main "aitem" but creating an extra "aitem" with null at last

```
{
  "root": {
    "aitem": {
      "name": "abc",
      "value": "123"
    },
    "bitem": [
      {
        "name": "bbbb",
        "value": "2222"
      },
      {
        "name": "BB",
        "value": "22"
      }
    ]
  },
  "aitem": [
    null
  ]
}

```

**Comment by mikefarah** [comment]
I already answered the double quoting question here: https://github.com/mikefarah/yq/discussions/1478

You can give yq multiple commands by using the `|` operator.

You can give multiple paths to a command using comma:

```
yq ' (.path.one, .path.two) |= [.] | <double quote command> | <something else> '
```

**Comment by ajit4sbi** [comment]
@mikefarah 
Thanks for the info, Could you please correct the command `yq -p=x -o=j '.aitem |= [.]' test.xml`
Below is the output, missed marking array at main "aitem" but creating an extra "aitem" with null at last
```
{
  "root": {
    "aitem": {
      "name": "abc",
      "value": "123"
    },
    "bitem": [
      {
        "name": "bbbb",
        "value": "2222"
      },
      {
        "name": "BB",
        "value": "22"
      }
    ]
  },
  "aitem": [
    null
  ]
}
```

**Comment by mikefarah** [comment]
Ah the real `aitem` is nested under 'root' - so the path should be `.root.aitem`

**Comment by ajit4sbi** [comment]
Hi @mikefarah 

Could you please correct the command

> yq ' (.content.exportDeclaration.lineItems.itemCodes, .content.exportDeclaration.tracking) |= [.] | (.. | select(tag == "!!str")) |= from_json ' -px -oj  JsonRequest.xml
> 
> Error: cannot index array with 'itemCodes' (strconv.ParseInt: parsing "itemCodes": invalid syntax)
> 

Defining below tree elements to be a mandatory array , But if multiple set of 'lineItems' with 'itemCodes' are there then its failing.
.content.exportDeclaration.lineItems.commodityCodes
.content.exportDeclaration.tracking

```
	<content>

		<description>ITEMS</description>
		<country>JAP</country>

		<exportDeclaration>
	
                        <lineItems>
				<number>1</number>
				<description>PRODUCT A121</description>
				<quantity>
					<value>1600</value>
					<unitOfMeasurement>PCS</unitOfMeasurement>
				</quantity>
				<itemCodes>
					<typeCode>outbound</typeCode>
					<value>82828282</value>
				</itemCodes>
				<exportReasonType>Permanent</exportReasonType>
				<manufacturerCountry>MY</manufacturerCountry>
			</lineItems>

			<lineItems>
				<number>2</number>
				<description>PRODUCT B121</description>
				<quantity>
					<value>1800</value>
					<unitOfMeasurement>PCS</unitOfMeasurement>
				</quantity>
				<itemCodes>
					<typeCode>outbound</typeCode>
					<value>84848484</value>
				</itemCodes>
				<exportReasonType>Permanent</exportReasonType>
				<manufacturerCountry>MY</manufacturerCountry>
			</lineItems>

			<tracking>
				<number>"123456789"</number>
				<date>"2023-02-28"</date>
			</tracking>

		</exportDeclaration>

	</content>
```
